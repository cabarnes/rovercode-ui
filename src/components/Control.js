import React from 'react';
import { connect } from 'react-redux';
import { Button, Grid, Tooltip } from '@material-ui/core';
import {
  PlayArrow,
  Replay,
  SkipNext,
  Stop,
} from '@material-ui/icons';
import { withStyles } from '@material-ui/core/styles';
import { hot } from 'react-hot-loader';
import { injectIntl } from 'react-intl';
import PropTypes from 'prop-types';

import {
  changeExecutionState as actionChangeExecutionState,
  EXECUTION_RUN,
  EXECUTION_STEP,
  EXECUTION_STOP,
  EXECUTION_RESET,
} from '@/actions/code';

const mapStateToProps = ({ code }) => ({ code });
const mapDispatchToProps = (dispatch) => ({
  changeExecutionState: (state) => dispatch(actionChangeExecutionState(state)),
});

const Control = ({
  code,
  changeExecutionState,
  isConnected,
  intl,
}) => {
  const stopTitle = intl.formatMessage({
    id: 'app.control.stop',
    description: 'Button label to stop execution of the code',
    defaultMessage: 'Stop',
  });

  const runTitle = intl.formatMessage({
    id: 'app.control.run',
    description: 'Button label to start execution of the code',
    defaultMessage: 'Run',
  });

  const stepTitle = intl.formatMessage({
    id: 'app.control.step',
    description: 'Button label to step one place forward in the code',
    defaultMessage: 'Step',
  });

  const resetTitle = intl.formatMessage({
    id: 'app.control.reset',
    description: 'Button label to reset to the beginning of the code',
    defaultMessage: 'Reset',
  });

  const RunButton = withStyles(() => ({
    root: {
      color: '#FFFFFF',
      width: '42px',
      height: '42px',
      minWidth: '0px',
      boxShadow: '0px 3px 3px #00000029',
      borderRadius: '4px',
      opacity: 1,
    },
  }))(Button);

  const StepButton = withStyles((theme) => ({
    root: {
      color: '#FFFFFF',
      backgroundColor: theme.palette.warning.main,
      '&:hover': {
        backgroundColor: theme.palette.warning.dark,
      },
      width: '24px',
      height: '24px',
      minWidth: '0px',
      boxShadow: '0px 1px 3px #00000057',
      borderRadius: '4px',
      opacity: 1,
    },
  }))(Button);

  const ResetButton = withStyles(() => ({
    root: {
      color: '#FFFFFF',
      width: '24px',
      height: '24px',
      minWidth: '0px',
      boxShadow: '0px 1px 3px #00000057',
      borderRadius: '4px',
      opacity: 1,
    },
  }))(Button);

  const StopButton = withStyles((theme) => ({
    root: {
      color: '#FFFFFF',
      backgroundColor: theme.palette.error.main,
      '&:hover': {
        backgroundColor: theme.palette.error.dark,
      },
      width: '42px',
      height: '42px',
      minWidth: '0px',
      boxShadow: '0px 3px 3px #00000029',
      borderRadius: '4px',
      opacity: 1,
    },
  }))(Button);

  return (
    <>
      <Grid container direction="row" justify="center" alignItems="center" spacing={1} style={{ minWidth: '150px' }}>
        {
        code.execution === EXECUTION_RUN ? (
          <Grid item>
            <Tooltip title={stopTitle}>
              <StopButton
                variant="contained"
                onClick={() => changeExecutionState(EXECUTION_STOP)}
              >
                <Stop />
              </StopButton>
            </Tooltip>
          </Grid>
        ) : (
          <>
            <Grid item>
              <Tooltip title={resetTitle}>
                <ResetButton
                  variant="contained"
                  color="secondary"
                  onClick={() => changeExecutionState(EXECUTION_RESET)}
                  disabled={!isConnected}
                >
                  <Replay />
                </ResetButton>
              </Tooltip>
            </Grid>
            <Grid item>
              <Tooltip title={runTitle}>
                <RunButton
                  variant="contained"
                  color="primary"
                  onClick={() => changeExecutionState(EXECUTION_RUN)}
                  disabled={!isConnected}
                >
                  <PlayArrow />
                </RunButton>
              </Tooltip>
            </Grid>
            <Grid item>
              <Tooltip title={stepTitle}>
                <StepButton
                  variant="contained"
                  onClick={() => changeExecutionState(EXECUTION_STEP)}
                  disabled={!isConnected}
                >
                  <SkipNext />
                </StepButton>
              </Tooltip>
            </Grid>
          </>
        )
      }
      </Grid>
    </>
  );
};

Control.defaultProps = {
  isConnected: false,
};

Control.propTypes = {
  code: PropTypes.shape({
    execution: PropTypes.number,
  }).isRequired,
  isConnected: PropTypes.bool,
  changeExecutionState: PropTypes.func.isRequired,
  intl: PropTypes.shape({
    formatMessage: PropTypes.func.isRequired,
  }).isRequired,
};

export default hot(module)(injectIntl(connect(mapStateToProps, mapDispatchToProps)(Control)));
