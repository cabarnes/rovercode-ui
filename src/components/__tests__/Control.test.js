import React from 'react';
import { Button } from '@material-ui/core';
import toJson from 'enzyme-to-json';
import configureStore from 'redux-mock-store';

import {
  changeExecutionState,
  EXECUTION_RUN,
  EXECUTION_STEP,
  EXECUTION_STOP,
  EXECUTION_RESET,
} from '@/actions/code';
import Control from '../Control';

describe('The Control component', () => {
  const mockStore = configureStore();
  let store;

  beforeEach(() => {
    store = mockStore({
      code: {
        execution: EXECUTION_STOP,
      },
    });
    store.dispatch = jest.fn();
  });

  test('renders on the page with no errors in stop state', () => {
    const wrapper = mountWithIntl(<Control store={store} isConnected />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('renders on the page with no errors in run state', () => {
    const runStore = mockStore({
      code: {
        execution: EXECUTION_RUN,
      },
    });
    const wrapper = mountWithIntl(<Control store={runStore} isConnected />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('dispatches an action to change execution state', () => {
    const wrapper = shallowWithIntl(<Control store={store} />).dive().dive().dive();

    wrapper.props().changeExecutionState(EXECUTION_RUN);

    expect(store.dispatch).toHaveBeenCalledWith(changeExecutionState(EXECUTION_RUN));
  });

  test('changes to run state on button press', () => {
    const wrapper = shallowWithIntl(<Control store={store} />).dive().dive().dive()
      .dive();

    wrapper.find('WithStyles(WithStyles(ForwardRef(Button)))').at(1).simulate('click');

    expect(store.dispatch).toHaveBeenCalledWith(changeExecutionState(EXECUTION_RUN));
  });

  test('changes to step state on button press', () => {
    const wrapper = shallowWithIntl(<Control store={store} />).dive().dive().dive()
      .dive();

    wrapper.find('WithStyles(WithStyles(ForwardRef(Button)))').at(2).simulate('click');

    expect(store.dispatch).toHaveBeenCalledWith(changeExecutionState(EXECUTION_STEP));
  });

  test('changes to reset state on button press', () => {
    const wrapper = shallowWithIntl(<Control store={store} />).dive().dive().dive()
      .dive();

    wrapper.find('WithStyles(WithStyles(ForwardRef(Button)))').at(0).simulate('click');

    expect(store.dispatch).toHaveBeenCalledWith(changeExecutionState(EXECUTION_RESET));
  });

  test('changes to stop state on button press', () => {
    const runStore = mockStore({
      code: {
        execution: EXECUTION_RUN,
      },
    });
    runStore.dispatch = jest.fn();
    const wrapper = shallowWithIntl(<Control store={runStore} />).dive().dive().dive()
      .dive();

    wrapper.find('WithStyles(WithStyles(ForwardRef(Button)))').at(0).simulate('click');

    expect(runStore.dispatch).toHaveBeenCalledWith(changeExecutionState(EXECUTION_STOP));
  });

  test('disables buttons when not connected', () => {
    const wrapper = shallowWithIntl(<Control store={store} />).dive().dive().dive()
      .dive();

    wrapper.find(Button).forEach((btn) => {
      expect(btn.prop('disabled')).toBe(true);
    });
  });
});
