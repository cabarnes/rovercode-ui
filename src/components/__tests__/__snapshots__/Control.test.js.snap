// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Control component renders on the page with no errors 1`] = `
<Connect(Control)
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [MockFunction],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Control
    changeExecutionState={[Function]}
    code={
      Object {
        "execution": 3,
      }
    }
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [MockFunction],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
  >
    <Button
      animated="vertical"
      as="button"
      color="green"
      onClick={[Function]}
      onMouseDown={[Function]}
      size="huge"
    >
      <Ref
        innerRef={
          Object {
            "current": <button
              class="ui green huge vertical animated button"
            >
              <div
                class="hidden content"
              >
                Run
              </div>
              <div
                class="visible content"
              >
                <i
                  aria-hidden="true"
                  class="play icon"
                />
              </div>
            </button>,
          }
        }
      >
        <RefFindNode
          innerRef={
            Object {
              "current": <button
                class="ui green huge vertical animated button"
              >
                <div
                  class="hidden content"
                >
                  Run
                </div>
                <div
                  class="visible content"
                >
                  <i
                    aria-hidden="true"
                    class="play icon"
                  />
                </div>
              </button>,
            }
          }
        >
          <button
            className="ui green huge vertical animated button"
            onClick={[Function]}
            onMouseDown={[Function]}
          >
            <ButtonContent
              hidden={true}
            >
              <div
                className="hidden content"
              >
                <FormattedMessage
                  defaultMessage="Run"
                  id="app.control.run"
                  values={Object {}}
                >
                  Run
                </FormattedMessage>
              </div>
            </ButtonContent>
            <ButtonContent
              visible={true}
            >
              <div
                className="visible content"
              >
                <Icon
                  as="i"
                  name="play"
                >
                  <i
                    aria-hidden="true"
                    className="play icon"
                    onClick={[Function]}
                  />
                </Icon>
              </div>
            </ButtonContent>
          </button>
        </RefFindNode>
      </Ref>
    </Button>
    <Button
      animated="vertical"
      as="button"
      color="yellow"
      onClick={[Function]}
      onMouseDown={[Function]}
      style={
        Object {
          "verticalAlign": "bottom",
        }
      }
    >
      <Ref
        innerRef={
          Object {
            "current": <button
              class="ui yellow vertical animated button"
              style="vertical-align: bottom;"
            >
              <div
                class="hidden content"
              >
                Step
              </div>
              <div
                class="visible content"
              >
                <i
                  aria-hidden="true"
                  class="step forward icon"
                />
              </div>
            </button>,
          }
        }
      >
        <RefFindNode
          innerRef={
            Object {
              "current": <button
                class="ui yellow vertical animated button"
                style="vertical-align: bottom;"
              >
                <div
                  class="hidden content"
                >
                  Step
                </div>
                <div
                  class="visible content"
                >
                  <i
                    aria-hidden="true"
                    class="step forward icon"
                  />
                </div>
              </button>,
            }
          }
        >
          <button
            className="ui yellow vertical animated button"
            onClick={[Function]}
            onMouseDown={[Function]}
            style={
              Object {
                "verticalAlign": "bottom",
              }
            }
          >
            <ButtonContent
              hidden={true}
            >
              <div
                className="hidden content"
              >
                <FormattedMessage
                  defaultMessage="Step"
                  id="app.control.step"
                  values={Object {}}
                >
                  Step
                </FormattedMessage>
              </div>
            </ButtonContent>
            <ButtonContent
              visible={true}
            >
              <div
                className="visible content"
              >
                <Icon
                  as="i"
                  name="step forward"
                >
                  <i
                    aria-hidden="true"
                    className="step forward icon"
                    onClick={[Function]}
                  />
                </Icon>
              </div>
            </ButtonContent>
          </button>
        </RefFindNode>
      </Ref>
    </Button>
    <Button
      animated="vertical"
      as="button"
      color="blue"
      onClick={[Function]}
      onMouseDown={[Function]}
      style={
        Object {
          "verticalAlign": "bottom",
        }
      }
    >
      <Ref
        innerRef={
          Object {
            "current": <button
              class="ui blue vertical animated button"
              style="vertical-align: bottom;"
            >
              <div
                class="hidden content"
              >
                Reset
              </div>
              <div
                class="visible content"
              >
                <i
                  aria-hidden="true"
                  class="repeat icon"
                />
              </div>
            </button>,
          }
        }
      >
        <RefFindNode
          innerRef={
            Object {
              "current": <button
                class="ui blue vertical animated button"
                style="vertical-align: bottom;"
              >
                <div
                  class="hidden content"
                >
                  Reset
                </div>
                <div
                  class="visible content"
                >
                  <i
                    aria-hidden="true"
                    class="repeat icon"
                  />
                </div>
              </button>,
            }
          }
        >
          <button
            className="ui blue vertical animated button"
            onClick={[Function]}
            onMouseDown={[Function]}
            style={
              Object {
                "verticalAlign": "bottom",
              }
            }
          >
            <ButtonContent
              hidden={true}
            >
              <div
                className="hidden content"
              >
                <FormattedMessage
                  defaultMessage="Reset"
                  id="app.control.reset"
                  values={Object {}}
                >
                  Reset
                </FormattedMessage>
              </div>
            </ButtonContent>
            <ButtonContent
              visible={true}
            >
              <div
                className="visible content"
              >
                <Icon
                  as="i"
                  name="repeat"
                >
                  <i
                    aria-hidden="true"
                    className="repeat icon"
                    onClick={[Function]}
                  />
                </Icon>
              </div>
            </ButtonContent>
          </button>
        </RefFindNode>
      </Ref>
    </Button>
  </Control>
</Connect(Control)>
`;
